captureformat.md

```diff
@@ -0,0 +1,120 @@
+
+All multi-byte values are LITTLE-ENDIAN.
+
+Base packet header
+------------------
+
+The USBPCAP\_BUFFER\_PACKET\_HEADER as defined in USBPcapBuffer.h:
+
+#pragma pack(1)
+typedef struct
+{
+    USHORT       headerLen; /\* This header length \*/
+    UINT64       irpId;     /\* I/O Request packet ID \*/
+    USBD\_STATUS  status;    /\* USB status code
+                               (on return from host controller) \*/
+    USHORT       function;  /\* URB Function \*/
+    UCHAR        info;      /\* I/O Request info \*/
+
+    USHORT       bus;       /\* bus (RootHub) number \*/
+    USHORT       device;    /\* device address \*/
+    UCHAR        endpoint;  /\* endpoint number and transfer direction \*/
+    UCHAR        transfer;  /\* transfer type \*/
+
+    UINT32       dataLength;/\* Data length \*/
+} USBPCAP\_BUFFER\_PACKET\_HEADER, \*PUSBPCAP\_BUFFER\_PACKET\_HEADER;
+
+*   headerLen (offset 0) describes the total length, in bytes, of the header (including all transfer-specific header data).
+*   irpId (offset 2) is merely a pointer to IRP casted to the UINT64. This value can be used to match the request with respons.
+*   status (offset 10) is valid only on return from host-controller. This field corrensponds to the Status member of [\_URB\_HEADER](http://msdn.microsoft.com/en-us/library/windows/hardware/ff540409(v=vs.85).aspx)
+*   function (offset 14) corrensponds to the Function member of [\_URB\_HEADER](http://msdn.microsoft.com/en-us/library/windows/hardware/ff540409(v=vs.85).aspx)  
+    
+*   info (offset 16) is descibed on the bit-field basis. Currently only the least significant bit (USBPCAP\_INFO\_PDO\_TO\_FDO) is defined: it is 0 when IRP goes from FDO to PDO, 1 the other way round. The remaining bits are reserved and must be set to 0.
+*   bus (offset 17) is the root hub identifier used to distingush between multiple root hubs.
+*   device (offset 19) is USB device number. This field is, contary to the USB specification, 16-bit because the Windows uses 16-bits value for that matter. Check DeviceAddress field of [USB\_NODE\_CONNECTION\_INFORMATION](http://msdn.microsoft.com/en-us/library/windows/hardware/ff540090(v=vs.85).aspx)
+*   endpoint (offset 21) is the endpoint number used on the USB bus (the MSB describes transfer direction)
+*   transfer (offset 22) determines the transfer type and thus the header type. See below for details.
+*   dataLength (offset 23) specifies the total length of transfer data to follow directly after the header (at offset headerLen).
+
+Transfer-specific headers
+-------------------------
+
+All transfer-specific headers inherit the USBPCAP\_BUFFER\_PACKET\_HEADER, so first there is the USBPCAP\_BUFFER\_PACKET\_HEADER, then (if any) additional transfer-specific header data and then the transfer data.
+
+### USBPCAP\_TRANSFER\_ISOCHRONOUS
+
+When transfer is equal to USBPCAP\_TRANSFER\_ISOCHRONOUS (0) the header type is USBPCAP\_BUFFER\_ISOCH\_HEADER  
+
+/\* Note about isochronous packets:
```

Comment:
These two structures should probably also be wrapped inside code block.

Reply:
Done !